lib/collection/src/operations/verification/query.rs

```diff
@@ -35,6 +35,54 @@ impl Query {
         }
         Ok(())
     }
+
+    /// Check that the query does not perform a fullscan based on the collection configuration.
+    async fn check_fullscan(
+        &self,
+        using: &str,
+        collection: &Collection,
+        strict_mode_config: &StrictModeConfig,
+    ) -> CollectionResult<()> {
+        // Check only applies on `search_allow_exact`
+        if strict_mode_config.search_allow_exact == Some(false) {
+            match &self {
+                Query::Fusion(_) | Query::OrderBy(_) | Query::Formula(_) | Query::Sample(_) => (),
+                Query::Vector(_) => {
+                    let config = collection.collection_config.read().await;
+
+                    // ignore sparse vectors
+                    let query_targets_sparse = config
+                        .params
+                        .sparse_vectors
+                        .as_ref()
+                        .is_some_and(|sparse| sparse.contains_key(using));
+                    if query_targets_sparse {
+                        // sparse vectors are always indexed
+                        return Ok(());
+                    }
+
+                    // check HNSW configuration for vector
+                    let vector_hnsw_config = &config
+                        .params
+                        .vectors
+                        .get_params(using)
+                        .and_then(|param| param.hnsw_config.as_ref());
+
+                    let vector_hnsw_m = vector_hnsw_config.and_then(|hnsw| hnsw.m);
+                    // TODO(strict-mode) check also payload_m if if there is a filter by tenant/principal
```

Comment:
Just FYI: we can't do releases without this todo implemented. Otherwise it is too restrictive